rest/route/host_spawn.go

```diff
@@ -1128,53 +1129,75 @@ func (h *hostRunScript) Factory() gimlet.RouteHandler {
 
 func (h *hostRunScript) Parse(ctx context.Context, r *http.Request) error {
 	var err error
-	h.hostID, err = validateID(gimlet.GetVars(r)["host_id"])
-	if err != nil {
-		return errors.Wrap(err, "invalid host id")
-	}
-
-	hostScript := model.APIHostScript{}
-	if err = util.ReadJSONInto(util.NewRequestReader(r), &hostScript); err != nil {
+	hostScriptOpts := model.APIHostScript{}
+	if err = util.ReadJSONInto(util.NewRequestReader(r), &hostScriptOpts); err != nil {
 		return errors.Wrap(err, "can't read host command from json")
 	}
-	h.script = hostScript.Script
+	h.script = hostScriptOpts.Script
+	h.hostIDs = hostScriptOpts.Hosts
 
 	return nil
 }
 
 func (h *hostRunScript) Run(ctx context.Context) gimlet.Responder {
-	user := MustHaveUser(ctx)
-	host, err := h.sc.FindHostByIdWithOwner(h.hostID, user)
-	if err != nil {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "can't get host"))
-	}
-	if host.Status != evergreen.HostRunning {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Errorf("can't run script on host with status '%s'", host.Status))
-	}
+	u := MustHaveUser(ctx)
+	type hostOutput struct {
+		hostID string
+		output string
+	}
+	hostsOutput := make(chan hostOutput, len(h.hostIDs))
+	wg := &sync.WaitGroup{}
+	wg.Add(len(h.hostIDs))
+	for _, hostID := range h.hostIDs {
+		go func(ctx context.Context, hostID string, u *user.DBUser) {
+			defer wg.Done()
+			host, err := h.sc.FindHostByIdWithOwner(hostID, u)
+			if err != nil {
+				hostsOutput <- hostOutput{hostID: hostID, output: errors.Wrap(err, "can't get host").Error()}
+				return
+			}
+			if host.Status != evergreen.HostRunning {
+				hostsOutput <- hostOutput{hostID: hostID, output: errors.Errorf("can't run script on host with status '%s'", host.Status).Error()}
+				return
+			}
 
-	if host.JasperCommunication() {
-		opts := &options.Create{
-			Args: []string{"bash", "-l", "-c", h.script},
-		}
-		output, err := host.RunJasperProcess(ctx, h.env, opts)
-		if err != nil {
-			return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "can't run script with Jasper"))
-		}
-		return gimlet.NewJSONResponse(model.APIHostScriptResponse{Output: output})
-	}
+			if host.JasperCommunication() {
+				opts := &options.Create{
+					Args: []string{"bash", "-l", "-c", h.script},
+				}
+				processOutput, err := host.RunJasperProcess(ctx, h.env, opts)
+				if err != nil {
+					hostsOutput <- hostOutput{hostID: hostID, output: errors.Wrap(err, "can't run script with Jasper").Error()}
+					return
+				}
+				hostsOutput <- hostOutput{hostID: hostID, output: strings.Join(processOutput, "\n")}
+				return
+			}
 
-	sshOptions, err := host.GetSSHOptions(h.env.Settings())
-	sshOptions = append(sshOptions, "-o", "LogLevel=quiet")
-	if err != nil {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "can't get ssh options"))
+			sshOptions, err := host.GetSSHOptions(h.env.Settings())
```

Comment:
I think this will be a bottleneck, enough so that we will not want as many goroutines as hosts. The jasper-over-rpc block is probably fine, but running hundreds of commands to different hosts concurrently over SSH may not work.

Reply:
In-person we decided it would likely be a bottleneck for Jasper as well, but it's no longer an issue given that the connections are not all open at once.